dds/DCPS/WriteDataContainer.cpp

```diff
@@ -1500,6 +1488,159 @@ WriteDataContainer::log_send_state_lists (OPENDDS_STRING description)
              instances_.size()));
 }
 
+void
+WriteDataContainer::set_deadline_period(const TimeDuration& deadline_period)
+{
+  // Call comes from DataWriterImpl_t which should arleady have the lock_.
+
+  // Reset the deadline timer if the period has changed.
+  if (deadline_period_ != deadline_period) {
+    if (deadline_period_ == TimeDuration::max_value) {
+      OPENDDS_ASSERT(deadline_map_.empty());
+
+      const MonotonicTimePoint deadline = MonotonicTimePoint::now() + deadline_period;
+
+      for (PublicationInstanceMapType::iterator iter = instances_.begin();
+           iter != instances_.end();
+           ++iter) {
+        iter->second->deadline_ = deadline;
+        deadline_map_.insert(std::make_pair(deadline, iter->second));
+      }
+
+      if (!deadline_map_.empty()) {
+        deadline_task_->schedule(deadline_period);
+      }
+    } else if (deadline_period == TimeDuration::max_value) {
+      if (!deadline_map_.empty()) {
+        deadline_task_->cancel();
+      }
+
+      deadline_map_.clear();
+    } else {
+      // TODO: Which is correct?  The first can cause immediate deadline misses?  The second corresponds to the existing behavior.
+      //const TimeDuration adjustment = deadline_period - deadline_period_;
+      const TimeDuration adjustment = deadline_period;
+      DeadlineMapType new_map;
+      for (PublicationInstanceMapType::iterator iter = instances_.begin();
+           iter != instances_.end();
+           ++iter) {
+        iter->second->deadline_ += adjustment;
+        new_map.insert(std::make_pair(iter->second->deadline_, iter->second));
+      }
+      std::swap(new_map, deadline_map_);
+
+      if (!deadline_map_.empty()) {
+        deadline_task_->cancel();
+        deadline_task_->schedule(deadline_map_.begin()->first - MonotonicTimePoint::now());
```

Comment:
I would maybe update the logic for calculating the next scheduled deadline similar between this line and how it's calculated above for a "new" deadline... i.e. have a temporary variable named deadline, just to make it clear what's going on here. Also be aware that sometimes canceling timers can take a while, so calculating "now" as close to function entry as possible is probably the desired behavior.

Reply:
In this case, it is better to defer now() to as late as possible.  If I use now() upon function entry, then the relative deadline to schedule will be larger and the timer will fire after the deadline.